Terminal.Gui/Windows/FileDialog2.cs

```diff
@@ -0,0 +1,2118 @@
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using NStack;
+using Terminal.Gui.Trees;
+using static System.Environment;
+using static Terminal.Gui.OpenDialog;
+
+namespace Terminal.Gui {
+
+	/// <summary>
+	/// Modal dialog for selecting files/directories.  Has auto-complete and expandable
+	/// navigation pane (Recent, Root drives etc).
+	/// </summary>
+	public class FileDialog2 : Dialog {
+
+		// TODO : expose these somehow for localization without compromising case/switch statements
+		private const string HeaderFilename = "Filename";
+		private const string HeaderSize = "Size";
+		private const string HeaderModified = "Modified";
+		private const string HeaderType = "Type";
+
+		/// <summary>
+		/// The maximum number of results that will be collected
+		/// when searching before stopping.
+		/// </summary>
+		/// <remarks>
+		/// This prevents performance issues e.g. when searching
+		/// root of file system for a common letter (e.g. 'e').
+		/// </remarks>
+		public int MaxSearchResults {get;set;} = 10000;
+
+		/// <summary>
+		/// True if the file/folder must exist already to be selected.
+		/// This prevents user from entering the name of something that
+		/// doesn't exist.  Defaults to false.
+		/// </summary>
+		public bool MustExist { get; set; }
+
+		private static char [] separators = new []
+		{
+			System.IO.Path.AltDirectorySeparatorChar,
+			System.IO.Path.DirectorySeparatorChar,
+		};
+
+		/// <summary>
+		/// Characters to prevent entry into <see cref="tbPath"/>.  Note that this is not using
+		/// <see cref="System.IO.Path.GetInvalidFileNameChars"/> because we do want to allow directory
+		/// separators, arrow keys etc.
+		/// </summary>
+		private static char [] badChars = new []
+		{
+			'"','<','>','|','*','?',
+		};
+
+
+		/// <summary>
+		/// The UI selected <see cref="AllowedType"/> from combo box. May be null.
+		/// </summary>
+		private AllowedType currentFilter;
+
+		private bool pushingState = false;
+
+		private FileDialogState state;
+		private object onlyOneSearchLock = new object ();
+		private bool disposed = false;
+		private TextFieldWithAppendAutocomplete tbPath;
+
+		private FileDialogSorter sorter;
+		private FileDialogHistory history;
+
+		private DataTable dtFiles;
+		private TableView tableView;
+		private TreeView<object> treeView;
+		private TileView splitContainer;
+		private Button btnOk;
+		private Button btnCancel;
+		private Button btnToggleSplitterCollapse;
+		private Label lblForward;
+		private Label lblBack;
+		private Label lblUp;
+		private Label lblFeedback;
+
+		private CollectionNavigator collectionNavigator = new CollectionNavigator ();
+
+		private CaptionedTextField tbFind;
+		private SpinnerLabel spinnerLabel;
+
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="FileDialog2"/> class.
+		/// </summary>
+		public FileDialog2 ()
+		{
+			const int okWidth = 6;
```

Comment:
Why this hackiness? Is there something missing from our computed layout system/button that leads you to using these hard coded values?

Reply:
It's very common to have dialogs with OK/Cancel. I wonder if we should push this down into `Dialog`?

Reply:
I will have a look and see if I can get rid of the const.  It is used with Anchor.AtEnd to position at lower right of window.

Could possibly be moved to `Dialog` but theres currently a lot of interaction with the button e.g. the arrow keys `NavigateIf` stuff and they are also positioned at the bottom right (i.e. not centered like would be with regular base class `Dialog`) and on the same row as other controls (Search, toggle tree and feedback).   

Ill do some coding and see where it goes to.